server/src/main/java/org/elasticsearch/rest/action/logs/RestLogsAction.java

```diff
@@ -0,0 +1,252 @@
+                    event.put("message", content.slice(offset, endOfEvent - offset).utf8ToString());
+                    event = createDlqDoc(event, e);
+                }
+                indexRequests.add(Requests.indexRequest(routeToDataStream(event)).opType(DocWriteRequest.OpType.CREATE).source(event));
+            }
+        }
+
+        return channel -> {
+            client.bulk(Requests.bulkRequest().add(indexRequests), new RestActionListener<>(channel) {
+                @Override
+                protected void processResponse(BulkResponse bulkItemResponses) throws Exception {
+                    if (bulkItemResponses.hasFailures() == false) {
+                        sendResponse(channel, RestStatus.ACCEPTED, b -> {});
+                        return;
+                    }
+                    BulkRequest retryBulk = Requests.bulkRequest();
+                    Arrays.stream(bulkItemResponses.getItems()).filter(BulkItemResponse::isFailed).forEach(failedRequest -> {
+                        IndexRequest originalRequest = indexRequests.get(failedRequest.getItemId());
+                        Map<String, Object> doc = originalRequest.sourceAsMap();
+                        Exception cause = failedRequest.getFailure().getCause();
+                        doc = createDlqDoc(doc, cause);
+                        retryBulk.add(Requests.indexRequest(routeToDataStream(doc)).opType(DocWriteRequest.OpType.CREATE).source(doc));
+                    });
+                    client.bulk(retryBulk, new RestActionListener<>(channel) {
+                        @Override
+                        protected void processResponse(BulkResponse bulkItemResponses) throws Exception {
+                            if (bulkItemResponses.hasFailures() == false) {
+                                sendResponse(channel, RestStatus.ACCEPTED, b -> {});
+                            } else {
+                                sendResponse(channel, RestStatus.INTERNAL_SERVER_ERROR, b -> {});
+                                logger.error(
+                                    "Failed to ingest logs: re-try batch has failures. First failure: {}",
+                                    Arrays.stream(bulkItemResponses.getItems())
+                                        .filter(BulkItemResponse::isFailed)
+                                        .findFirst()
+                                        .map(BulkItemResponse::getFailureMessage)
+                                        .orElse(null)
+                                );
+                            }
+                        }
+                    });
+                }
+            });
+        };
+    }
+
+    private Map<String, Object> createDlqDoc(Map<String, Object> doc, Exception cause) {
+        Map<String, Object> dlqDoc = new HashMap<>();
+        dlqDoc.put("@timestamp", Instant.now().toString());
+        addPath(dlqDoc, "event.original", doc);
```

Comment:
The problem with this and the `error.*` is on what happens, if these fields already existed in the event? Will these be overwritten? Ideally we would store this data in a predefined structure in the document that can't conflict, something more "internal". Maybe we could start it with `_` or `.` like `_internal.event.orginal` and same for error. This has some disadvantages that it makes querying the dlq data harder but some magic on the ES side could help here assuming this is a special "index".

Reply:
> Will these be overwritten?

No, the original event is stored under event.original so that there's no opportunity for a conflict:

```json
{
  "error": {
    "type": "mapper_parsing_exception",
    "message": "..."
  },
  "event": {
    "original": {
      "@timestamp": "...",
      "message": "Hello World"
    }
  }
}
```

It means that the original event is not on the top level anymore but I think that's a fine tradeoff.